File: common/python/example_poll.py
# Copyright 2016 Zurich Instruments AG

"""
Zurich Instruments LabOne Python API Example

Demonstrate how to obtain demodulator data using ziDAQServer's blocking
(synchronous) poll() command.

This example demonstrates that whilst poll() does indeed block for the specified
recording duration, it will not only return the data during the recording
duration, but also data accumulated since subscribing (before poll was
called). In other words, subscribed data is buffered by the data server and API
(for a limited time) and this buffered data will also be returned by poll().

Connect to the device specified by device_id and obtain demodulator data using
ziDAQServer's blocking (synchronous) poll() command.

Requirements:
    * LabOne Version >= 20.02
    * Instruments:
        1 x Instrument with demodulators
    * signal output 1 connected to signal input 1 with a BNC cable.

Usage:
    example_poll.py [options] <device_id>
    example_poll.py -h | --help

Arguments:
    <device_id>  The ID of the device [device_type: .*LI|.*IA|.*IS]

Options:
    -h --help                 Show this screen.
    -a --amplitude AMPLITUDE  The amplitude to set on the signal output. [default: 0.5]
    --no-plot                 Hide plot of the recorded data.

Raises:
    Exception     If the specified devices do not match the requirements.
    RuntimeError  If the devices is not "discoverable" from the API.

See the "LabOne Programming Manual" for further help, available:
    - On Windows via the Start-Menu:
    Programs -> Zurich Instruments -> Documentation
    - On Linux in the LabOne .tar.gz archive in the "Documentation"
    sub-folder.
"""

import time
import numpy as np
import zhinst.utils
import matplotlib.pyplot as plt


def run_example(device_id: str, amplitude: float = 0.5, plot: bool = True):
    """run the example."""

    apilevel_example = 6  # The API level supported by this example.
    # Call a zhinst utility function that returns:
    # - an API session `daq` in order to communicate with devices via the data server.
    # - the device ID string that specifies the device branch in the server's node hierarchy.
    # - the device's discovery properties.
    err_msg = "This example only supports instruments with demodulators."
    (daq, device, props) = zhinst.utils.create_api_session(
        device_id,
        apilevel_example,
        required_devtype=".*LI|.*IA|.*IS",
        required_err_msg=err_msg,
    )
    zhinst.utils.api_server_version_check(daq)

    # Create a base configuration: Disable all available outputs, awgs, demods, scopes,...
    zhinst.utils.disable_everything(daq, device)

    # Now configure the instrument for this experiment. The following channels
    # and indices work on all device configurations. The values below may be
    # changed if the instrument has multiple input/output channels and/or either
    # the Multifrequency or Multidemodulator options installed.
    out_channel = 0
    out_mixer_channel = zhinst.utils.default_output_mixer_channel(props)
    in_channel = 0
    demod_index = 0
    osc_index = 0
    demod_rate = 1e3
    time_constant = 0.01
    frequency = 400e3
    exp_setting = [
        ["/%s/sigins/%d/ac" % (device, in_channel), 0],
        ["/%s/sigins/%d/range" % (device, in_channel), 2 * amplitude],
        ["/%s/demods/%d/enable" % (device, demod_index), 1],
        ["/%s/demods/%d/rate" % (device, demod_index), demod_rate],
        ["/%s/demods/%d/adcselect" % (device, demod_index), in_channel],
        ["/%s/demods/%d/order" % (device, demod_index), 4],
        ["/%s/demods/%d/timeconstant" % (device, demod_index), time_constant],
        ["/%s/demods/%d/oscselect" % (device, demod_index), osc_index],
        ["/%s/demods/%d/harmonic" % (device, demod_index), 1],
        ["/%s/oscs/%d/freq" % (device, osc_index), frequency],
        ["/%s/sigouts/%d/on" % (device, out_channel), 1],
        ["/%s/sigouts/%d/enables/%d" % (device, out_channel, out_mixer_channel), 1],
        ["/%s/sigouts/%d/range" % (device, out_channel), 1],
        [
            "/%s/sigouts/%d/amplitudes/%d" % (device, out_channel, out_mixer_channel),
            amplitude,
        ],
    ]
    daq.set(exp_setting)

    # Unsubscribe any streaming data.
    daq.unsubscribe("*")

    # Wait for the demodulator filter to settle.
    time.sleep(10 * time_constant)

    # Perform a global synchronisation between the device and the data server:
    # Ensure that 1. the settings have taken effect on the device before issuing
    # the poll() command and 2. clear the API's data buffers. Note: the sync()
    # must be issued after waiting for the demodulator filter to settle above.
    daq.sync()

    # Subscribe to the demodulator's sample node path.
    path = "/%s/demods/%d/sample" % (device, demod_index)
    daq.subscribe(path)

    # Sleep for demonstration purposes: Allow data to accumulate in the data
    # server's buffers for one second: poll() will not only return the data
    # accumulated during the specified poll_length, but also for data
    # accumulated since the subscribe() or the previous poll.
    sleep_length = 1.0
    # For demonstration only: We could, for example, be processing the data
    # returned from a previous poll().
    time.sleep(sleep_length)

    # Poll the subscribed data from the data server. Poll will block and record
    # for poll_length seconds.
    poll_length = 0.1  # [s]
    poll_timeout = 500  # [ms]
    poll_flags = 0
    poll_return_flat_dict = True
    data = daq.poll(poll_length, poll_timeout, poll_flags, poll_return_flat_dict)

    # Unsubscribe from all paths.
    daq.unsubscribe("*")

    # Check the dictionary returned is non-empty
    assert (
        data
    ), "poll() returned an empty data dictionary, did you subscribe to any paths?"

    # The data returned is a dictionary of dictionaries that reflects the node's path.
    # Note, the data could be empty if no data had arrived, e.g., if the demods
    # were disabled or had demodulator rate 0.
    assert path in data, "The data dictionary returned by poll has no key `%s`." % path

    # Access the demodulator sample using the node's path.
    sample = data[path]

    # Let's check how many seconds of demodulator data were returned by poll.
    # First, get the sampling rate of the device's ADCs, the device clockbase...
    clockbase = float(daq.getInt("/%s/clockbase" % device))
    # ... and use it to convert sample timestamp ticks to seconds:
    dt_seconds = (sample["timestamp"][-1] - sample["timestamp"][0]) / clockbase
    print(f"poll() returned {dt_seconds:.3f} seconds of demodulator data.")
    tol_percent = 10
    dt_seconds_expected = sleep_length + poll_length
    assert (
        dt_seconds - dt_seconds_expected
    ) / dt_seconds_expected * 100 < tol_percent, (
        "Duration of demod data returned by poll() (%.3f s) differs " % dt_seconds
        + "from the expected duration (%.3f s) by more than %0.2f %%."
        % (dt_seconds_expected, tol_percent)
    )

    # Calculate the demodulator's magnitude and phase and add them to the dict.
    # The phase is returned in units of radian. We can use a keyword for the numpy
    # angle function to return degree units instead: np.angle( ... , deg=True)
    sample["R"] = np.abs(sample["x"] + 1j * sample["y"])
    sample["phi"] = np.angle(sample["x"] + 1j * sample["y"])
    print(f"Average measured RMS amplitude is {np.mean(sample['R']):.3e} V.")

    if plot:

        # Convert timestamps from ticks to seconds via clockbase.
        t = (sample["timestamp"] - sample["timestamp"][0]) / clockbase

        # Create plot
        _, axis = plt.subplots()
        axis.grid(True)
        axis.plot(t, sample["R"])
        axis.set_title("Demodulator data")
        axis.set_xlabel("Time (s)")
        axis.set_ylabel("R (V)")
        mean_r = np.mean(sample["R"])
        axis.axis([t[0], t[-1], 0.99 * mean_r, 1.01 * mean_r])

        plt.show()


if __name__ == "__main__":
    import sys
    from pathlib import Path

    cli_util_path = Path(__file__).resolve().parent / "../../utils/python"
    sys.path.insert(0, str(cli_util_path))
    cli_utils = __import__("cli_utils")
    cli_utils.run_commandline(run_example, __doc__)
    sys.path.remove(str(cli_util_path))

File: common/matlab/example_poll.m
function sample = example_poll(device_id, varargin)
% EXAMPLE_POLL Record demodulator data using ziDAQServer's synchronous poll function
%
% USAGE DATA = EXAMPLE_POLL(DEVICE_ID)
%
% Poll demodulator sample data from the device specified by DEVICE_ID using
% ziDAQServer's poll method. DEVICE_ID should be a string, e.g., 'dev1000' or
% 'uhf-dev1000'.
%
% ziDAQServer's poll method allows the user to obtain ('poll') demodulator
% data. Data can be obtained continuously in a loop. If asynchronous data
% recording is necessary please see example_record_async.m which uses the
% ziDAQRecord module.
%
% NOTE Additional configuration: Connect signal output 1 to signal input 1
% with a BNC cable.
%
% NOTE Please ensure that the ziDAQ folders 'Driver' and 'Utils' are in your
% Matlab path. To do this (temporarily) for one Matlab session please navigate
% to the ziDAQ base folder containing the 'Driver', 'Examples' and 'Utils'
% subfolders and run the Matlab function ziAddPath().
% >>> ziAddPath;
%
% Use either of the commands:
% >>> help ziDAQ
% >>> doc ziDAQ
% in the Matlab command window to obtain help on all available ziDAQ commands.
%
% Copyright 2008-2018 Zurich Instruments AG

clear ziDAQ;

if ~exist('device_id', 'var')
    error(['No value for device_id specified. The first argument to the ' ...
           'example should be the device ID on which to run the example, ' ...
           'e.g. ''dev1000'' or ''uhf-dev1000''.'])
end

% Check the ziDAQ MEX (DLL) and Utility functions can be found in Matlab's path.
if ~(exist('ziDAQ') == 3) && ~(exist('ziCreateAPISession', 'file') == 2)
    fprintf('Failed to either find the ziDAQ mex file or ziDevices() utility.\n')
    fprintf('Please configure your path using the ziDAQ function ziAddPath().\n')
    fprintf('This can be found in the API subfolder of your LabOne installation.\n');
    fprintf('On Windows this is typically:\n');
    fprintf('C:\\Program Files\\Zurich Instruments\\LabOne\\API\\MATLAB2012\\\n');
    return
end

% The API level supported by this example.
apilevel_example = 6;
% Create an API session; connect to the correct Data Server for the device.
[device, props] = ziCreateAPISession(device_id, apilevel_example);
ziApiServerVersionCheck();

branches = ziDAQ('listNodes', ['/' device ], 0);
if ~any(strcmpi([branches], 'DEMODS'))
  sample = [];
  fprintf('\nThis example requires lock-in functionality which is not available on %s.\n', device);
  return
end

%% Define parameters relevant to this example. Default values specified by the
% inputParser below are overwritten if specified as name-value pairs via the
% `varargin` input argument.
p = inputParser;
isnonneg = @(x) isnumeric(x) && isscalar(x) && (x > 0);
% The length of time we'll record data (synchronously) in the first poll [s].
p.addParamValue('poll_duration', 1.0, isnonneg);
% The length of time to accumulate subscribed data (by sleeping) before polling a second time [s].
p.addParamValue('sleep_duration', 1.0, isnonneg);
% The signal output mixer amplitude, [V].
p.addParamValue('amplitude', 0.5, @isnumeric);
p.parse(varargin{:});
poll_duration = p.Results.poll_duration;
sleep_duration = p.Results.sleep_duration;
amplitude = p.Results.amplitude;

%% Define some other helper parameters.
demod_c = '0'; % Demod channel, 0-based indexing for paths on the device.
demod_idx = str2double(demod_c) + 1; % 1-based indexing, to access the data.
out_c = '0'; % signal output channel
% Get the value of the instrument's default Signal Output mixer channel.
out_mixer_c = num2str(ziGetDefaultSigoutMixerChannel(props, str2num(out_c)));
in_c = '0'; % signal input channel
osc_c = '0'; % oscillator
time_constant = 0.001; % [s]
demod_rate = 2e3;

% Create a base configuration: Disable all available outputs, awgs,
% demods, scopes,...
ziDisableEverything(device);

%% Configure the device ready for this experiment.
ziDAQ('setInt', ['/' device '/sigins/' in_c '/imp50'], 0);
ziDAQ('setInt', ['/' device '/sigins/' in_c '/ac'], 0);
ziDAQ('setDouble', ['/' device '/sigins/' in_c '/range'], 2.0*amplitude);
ziDAQ('setInt', ['/' device '/sigouts/' out_c '/on'], 1);
ziDAQ('setDouble', ['/' device '/sigouts/' out_c '/range'], 1);
ziDAQ('setDouble', ['/' device '/sigouts/' out_c '/amplitudes/*'], 0);
ziDAQ('setDouble', ['/' device '/sigouts/' out_c '/amplitudes/' out_mixer_c], amplitude);
ziDAQ('setDouble', ['/' device '/sigouts/' out_c '/enables/' out_mixer_c], 1);
if strfind(props.devicetype, 'HF2')
    ziDAQ('setInt', ['/' device '/sigins/' in_c '/diff'], 0);
    ziDAQ('setInt', ['/' device '/sigouts/' out_c '/add'], 0);
end
ziDAQ('setDouble', ['/' device '/demods/*/phaseshift'], 0);
ziDAQ('setInt', ['/' device '/demods/*/order'], 8);
ziDAQ('setDouble', ['/' device '/demods/' demod_c '/rate'], demod_rate);
ziDAQ('setInt', ['/' device '/demods/' demod_c '/harmonic'], 1);
ziDAQ('setInt', ['/' device '/demods/' demod_c '/enable'], 1);
ziDAQ('setInt', ['/' device '/demods/*/oscselect'], str2double(osc_c));
ziDAQ('setInt', ['/' device '/demods/*/adcselect'], str2double(in_c));
ziDAQ('setDouble', ['/' device '/demods/*/timeconstant'], time_constant);
ziDAQ('setDouble', ['/' device '/oscs/' osc_c '/freq'], 30e5); % [Hz]

% Unsubscribe all streaming data.
ziDAQ('unsubscribe', '*');

% Pause to get a settled lowpass filter.
pause(10*time_constant);

% Perform a global synchronisation between the device and the data server:
% Ensure that the settings have taken effect on the device before issuing the
% ``poll`` command and clear the API's data buffers to remove any old
% data. Note: ``sync`` must be issued after waiting for the demodulator filter
% to settle above.
ziDAQ('sync');

% Subscribe to the demodulator sample.
ziDAQ('subscribe', ['/' device '/demods/' demod_c '/sample']);

% Poll data for poll_duration seconds.
poll_timeout = 500;
data = ziDAQ('poll', poll_duration, poll_timeout);

figure(1); clf;
grid on; box on; hold on;
if ziCheckPathInData(data, ['/' device '/demods/' demod_c '/sample']);
    sample = data.(device).demods(demod_idx).sample;
    t0 = sample.timestamp(1);
    plot_sample(sample, device, t0, 'b-');
else
    sample = [];
end

% Wait 2 seconds and poll data for 0.01 seconds.
% NOTE we get all the data from the buffer since the last poll command (much
% more data than 0.01 seconds)!
pause(sleep_duration);
poll_timeout = 0;
data2 = ziDAQ('poll', 0.01, poll_timeout);
if ziCheckPathInData(data2, ['/' device '/demods/' demod_c '/sample']);
    sample2 = data2.(device).demods(demod_idx).sample;
    if ~exist('t0', 'var')
        t0 = sample2.timestamp(1);
    end
    plot_sample(sample2, device, t0, 'k-');
    legend('poll 1', 'poll 2');
end

% Unsubscribe from all paths.
ziDAQ('unsubscribe', '*');

end

function plot_sample(sample, device, t0, style)
if sample.time.dataloss
    fprintf('Warning: Sample loss detected.');
end
r = sqrt(sample.x.^2 + sample.y.^2);
% convert timestamps from ticks to seconds via the device's clockbase
% (the ADC's sampling rate), specify reference start time via t0.
clockbase = double(ziDAQ('getInt', ['/' device '/clockbase']));
t = (double(sample.timestamp) - double(t0))/clockbase;
fprintf('Poll returned %.3f seconds of data\n', t(end)-t(1));
plot(t, r, style);
end
